Lib/test/test_typing.py

```diff
@@ -2535,6 +2535,128 @@ def meth(x): ...
         with self.assertRaises(TypeError):
             isinstance(C(), BadPG)
 
+    def test_protocols_isinstance_simple_properties(self):
+        T = TypeVar('T')
+
+        @runtime_checkable
+        class P(Protocol):
+            @property
+            def attr(self): ...
+
+        @runtime_checkable
+        class P1(Protocol):
+            attr: int
+
+        @runtime_checkable
+        class PG(Protocol[T]):
+            @property
+            def attr(self): ...
+
+        @runtime_checkable
+        class PG1(Protocol[T]):
+            attr: T
+
+        class BadP(Protocol):
+            @property
+            def attr(self): ...
+
+        class BadP1(Protocol):
+            attr: int
+
+        class BadPG(Protocol[T]):
+            @property
+            def attr(self): ...
+
+        class BadPG1(Protocol[T]):
+            attr: T
+
+        class C:
+            @property
+            def attr(self):
+                return 42
+
+        self.assertEqual(C().attr, 42)
+        self.assertIsInstance(C(), P)
+        self.assertIsInstance(C(), P1)
+        self.assertIsInstance(C(), PG)
+        self.assertIsInstance(C(), PG1)
+        with self.assertRaises(TypeError):
+            isinstance(C(), PG[T])
+        with self.assertRaises(TypeError):
+            isinstance(C(), PG[C])
+        with self.assertRaises(TypeError):
+            isinstance(C(), PG1[T])
+        with self.assertRaises(TypeError):
+            isinstance(C(), PG1[C])
+        with self.assertRaises(TypeError):
+            isinstance(C(), BadP)
+        with self.assertRaises(TypeError):
+            isinstance(C(), BadP1)
+        with self.assertRaises(TypeError):
+            isinstance(C(), BadPG)
+        with self.assertRaises(TypeError):
+            isinstance(C(), BadPG1)
+
+    def test_protocols_isinstance_dynamic_properties(self):
+        T = TypeVar('T')
+
+        @runtime_checkable
+        class P(Protocol):
+            @property
+            def attr(self): ...
+
+        @runtime_checkable
+        class P1(Protocol):
+            attr: int
+
+        @runtime_checkable
+        class PG(Protocol[T]):
+            @property
+            def attr(self): ...
+
+        @runtime_checkable
+        class PG1(Protocol[T]):
+            attr: T
+
+        class C:
+            X = True
+            @property
+            def attr(self):
+                if self.X:
+                    return 42
+                raise AttributeError
+
+        inst = C()
+        C.X = False
+
+        with self.assertRaises(AttributeError):
+            C().attr
+        self.assertNotIsInstance(C(), P)
```

Comment:
We're probably constrained by backward compatibility now but I'm not convinced this is the right behavior. To a type checker, `C` would match the `P` protocol because it has the right attribute, and ideally `@runtime_checkable` should reflect that behavior.

Reply:
I disagree; I think the current behaviour is correct. For me, the important invariant is that the following is always type-safe:

```python
inst: object = C()
if isinstance(inst, P):
    bar = inst.attr
```

If we changed the behaviour here, then the `isinstance()` check above would not be adequate to ensure that accessing `inst.attr` is type-safe.